File: frontend/static/widgets/free_form_content/iframe.mjs
import { FreeFormContent } from './free_form_content.mjs'
import { ContentAndCaption } from '../containers/content_and_caption.mjs'
import { Caption } from '../caption.mjs'

/**
 * A piece of {@link FreeFormContent} which displays the embedded content of a link along with a potential caption.
 *
 * @augments FreeFormContent
 */
export class Iframe extends FreeFormContent {
  /**
   * @param {int} id the content's ID
   * @param {URL} url the URL to embed in the iFrame
   * @param {?Caption} caption the iframe's caption
   */
  constructor ({ id, url, caption }) {
    super({ id })
    this.url = url
    this.caption = caption
  }

  /**
   * Deserialize the Link from its JSON API representation.
   *
   * @param obj
   * @returns {Link}
   */
  static fromJSON (obj) {
    return new Iframe({
      id: obj.id,
      url: obj.url,
      caption: Caption.maybeFromJSON(obj.caption)
    })
  }

  build () {
    const iframeContent = document.createElement('iframe')
    iframeContent.src = this.url
    return new ContentAndCaption({
      content: iframeContent,
      caption: this.caption
    })
  }

  className () {
    return 'iframe-content'
  }
}

File: backend/src/server/free_form_content/iframe.py
from datetime import datetime
from typing import Optional

from server.free_form_content.caption import Caption
from server.free_form_content.captioned_content import CaptionedContent
from server.util import combine


class IFrameContent(CaptionedContent):
    """A link to be embedded in an iframe on the board along with a caption."""

    def __init__(
        self,
        url: str,
        caption: Optional[Caption],
        streams: list[int],
        content_id: Optional[int] = None,
        posted: Optional[datetime] = None,
    ):
        super().__init__(caption, streams, content_id, posted)
        self.url = url
        self.caption = caption

    def type(self) -> str:
        return "iframe_content"

    def to_db_json(self) -> dict:
        return combine(super().to_db_json(), {"url": self.url})
